lib/jsonrpc_client.ml

```diff
@@ -17,49 +17,103 @@
 module D = Debug.Make(struct let name = "jsonrpc_client" end)
 open D
 
-let input_json_object fin =
-	let buf = Buffer.create 1024 in
-	let brace_cnt = ref 0 in
-	let in_string = ref false in
-	let last_char () = Buffer.nth buf (Buffer.length buf - 1) in
-	let rec get () =
-		let c = input_char fin in
+exception Timeout
+exception Read_error
+
+let json_rpc_max_len = ref 65536 (* Arbitrary maximum length of RPC response *)
+let json_rpc_read_timeout = ref "60000000000"
+let json_rpc_write_timeout = ref "60000000000"
+
+let to_s  s = (Int64.to_float s) *. 1e-9
+
+(* Read the entire contents of the fd, of unknown length *)
+let timeout_read fd timeout =
+	let buf = Buffer.create !json_rpc_max_len in
+	let read_start = Mtime_clock.counter () in
+	let get_total_used_time () = Mtime.Span.to_uint64_ns (Mtime_clock.count read_start) in
+	let rec inner max_time max_bytes =
+		let (ready_to_read, _, _) = try Unix.select [fd] [] [] (to_s max_time) with
+			(* in case the unix.select call fails in situation like interrupt *)
+			| _ -> [], [], []
+		in
+		(* This is not accurate the calculate time just for the select part. However, we
+		 * think the read time will be minor comparing to the scale of tens of seconds.
+		 * the current style will be much concise in code. *)
+		let remain_time = 
+			let used_time = get_total_used_time () in
+			Int64.sub timeout used_time
+		in
+		if remain_time < 0L then 
 		begin
-			match c with
-			| '{' when not !in_string -> brace_cnt := !brace_cnt + 1
-			| '}' when not !in_string -> brace_cnt := !brace_cnt - 1
-			| '"' when !in_string && (last_char () <> '\\') -> in_string := false
-			| '"' when not !in_string -> in_string := true
-			| _ -> ()
+			debug "Timeout after read %d" (Buffer.length buf);
+			raise Timeout
 		end;
-		Buffer.add_char buf c;
-		if !brace_cnt > 0 then
-			get ()
+		if List.mem fd ready_to_read then
+		begin
+			let bytes = Bytes.make 4096 '\000' in
+			match Unix.read fd bytes 0 4096 with
+				| 0 -> Buffer.contents buf (* EOF *)
+				| n ->
+					if n > max_bytes then
+					begin
+						debug "exceeding maximum read limit %d, clear buffer" !json_rpc_max_len;
+						Buffer.clear buf;
+						raise Read_error
+					end
+					else
+					begin
+						Buffer.add_subbytes buf bytes 0 n;
+						inner remain_time (max_bytes - n)
+					end
+				| exception Unix.Unix_error(err,_,_) when err = Unix.EAGAIN || err = Unix.EWOULDBLOCK ->
```

Comment:
`| exception Unix.Unix_error(Unix.EAGAIN | Unix.EWOULDBLOCK,_,_)`  -> would be better

Reply:
`EINTR`?